mapie/classification.py

```diff
@@ -658,95 +859,147 @@ def predict(
+                        np.quantile(
+                            self.conformity_scores_,
+                            ((n + 1) * (1 - _alpha)) / n,
+                            interpolation="higher"
+                        ) for _alpha in alpha_
+                    ])
+                else:
+                    self.quantiles_ = (n + 1) * (1 - alpha_)
+
+            # Build prediction sets
             if self.method == "score":
-                prediction_sets = np.stack(
-                    [
-                        y_pred_proba > 1 - quantile
-                        for quantile in self.quantiles_
-                    ],
-                    axis=2,
-                )
-            elif self.method in ["cumulated_score", "naive"]:
-                # sort labels by decreasing probability
-                index_sorted = np.fliplr(np.argsort(y_pred_proba, axis=1))
-
-                # sort probabilities by decreasing order
-                y_pred_proba_sorted = np.take_along_axis(
-                    y_pred_proba, index_sorted, axis=1
-                )
-
-                # get sorted cumulated score
-                y_pred_proba_sorted_cumsum = np.cumsum(
-                    y_pred_proba_sorted, axis=1
-                )
+                if (cv == "prefit") or (agg_scores == "mean"):
+                    prediction_sets = np.stack(
+                        [
+                            y_pred_proba > 1 - quantile
+                            for quantile in self.quantiles_
+                        ],
+                        axis=2,
+                    )
+                else:
+                    y_comp = (
+                        1 - y_pred_proba < self.conformity_scores_[:, None]
+                    ).sum(axis=0)
+                    prediction_sets = np.stack(
+                        [
+                            y_comp > (n + 1 - quantile_)
+                            for quantile_ in self.quantiles_
+                        ], axis=2
+                    )
 
-                # get cumulated score at their original position
-                y_pred_proba_cumsum = np.take_along_axis(
-                    y_pred_proba_sorted_cumsum,
-                    np.argsort(index_sorted),
-                    axis=1
-                )
-                # get index of the last included label
-                y_pred_index_last = self._get_last_index_included(
-                    y_pred_proba_cumsum,
-                    include_last_label
-                )
-                # get the probability of the last included label
-                y_pred_proba_last = np.stack(
-                    [
-                        np.take_along_axis(
-                            y_pred_proba,
-                            y_pred_index_last[:, iq].reshape(-1, 1),
-                            axis=1
+            elif self.method in ["cumulated_score", "naive"]:
+                if (cv == "prefit") or (agg_scores in ["mean"]):
```

Comment:
This branching structure begins to grow very complicated. I wonder if we shouldn't draw it and refactorize some things.

Reply:
Agreed, let's discuss about this once this PR is validated.